### ISSUE TITLE
[cryptolib] Add NULL padding for RSA encrypt/decrypt

### ISSUE THREAD

**Issue by vsukhoml** [lowRISC/opentitan#21931]
### Description

Ti50 firmware have to support a use case where padding for RSA encrypt is done externally. Current RSA implementation doesn't support this.
For this NULL padding RSA encrypt implementation shall:
1) Pad input to RSA length as needed with zeroes
2) Check that input value is within valid range (less than N), remove leading zeroes to fit into RSA size

@moidx , @jadephilipoom , @cfrantz

**Comment by jadephilipoom** [comment]
This is a pretty dangerous thing to support. RSA padding is security-critical. Is there any way you can support this use-case without a direct option in the cryptolib? And do you need to support any padding other than OAEP?

I'd be okay with supporting a "pre-padded" option, where you assert that the input is already (for example) OAEP-padded, and cryptolib checks that this is true using the same method as we'd use to check padding in decryption. That would allow the use-case you describe, but it would also avoid exposing an option that completely skips the padding step.

**Comment by vsukhoml** [comment]
This is a legacy stuff, which is unfortunately used on the very critical path used now. The existing authentication method use CKM_MD5_RSA_PKCS mechanism, which beyond using legacy hash,  seems to differ from TPM_ALG_RSAES (PKCS1) supported by TPM directly, so it is done outside to limit changes in TPM2 and crypto and not to bring MD5.  From https://groups.oasis-open.org/higherlogic/ws/public/download/50264/CKM_RSA_PKCS_FIPS_186_4_v02.pdf/latest:
```
It differs from the standard PKCS#1 OAEP RSA encryption mechanism in that the plaintext is wrapped in a TPM_BOUND_DATA structure before being submitted to the encryption process and that all of the values of the parameters that are passed to a standard CKM_RSA_PKCS_OAEP operation are fixed. On encryption, the version field of the TPM_BOUND_DATA structure must contain 0x01, 0x01, 0x00, 0x00. On decryption, any structure of the
form 0x01, 0x01, 0xXX, 0xYY may be accepted.
```

https://chromium.googlesource.com/chromiumos/third_party/tpm2/+/569c3c58dc69d5c8628f3c329937c136be38df3f

Besides OAEP,  RSAES-PKCS1-v1_5 from https://datatracker.ietf.org/doc/html/rfc3447  is needed for the similar kind of legacy stuff.

**Comment by jadephilipoom** [comment]
> This is a legacy stuff, which is unfortunately used on the very critical path used now. The existing authentication method use CKM_MD5_RSA_PKCS mechanism, which beyond using legacy hash, seems to differ from TPM_ALG_RSAES (PKCS1) supported by TPM directly, so it is done outside to limit changes in TPM2 and crypto and not to bring MD5. From https://groups.oasis-open.org/higherlogic/ws/public/download/50264/CKM_RSA_PKCS_FIPS_186_4_v02.pdf/latest

The link goes to a "working draft" from more than 10 years ago that mentions the name but doesn't say what the algorithm actually does. Can you provide a little more context for what it is and where it is needed, if possible? The link also references TPM 1.1, but my understanding was that you needed TPM2, no? Using MD5 in 2024 seems like a surprising choice in any context, although I appreciate the need to support legacy setups.

Do I understand correctly that:
- there is some legacy system you cannot change but need to support
- it is separate from TPM and has its own requirements
- it uses RSA encryption with MD5 and PKCS#1 v1.5 padding and doesn't support any other option, and
- it uses this RSA encryption as an authentication mechanism?

The last point is additionally surprising to me, because I wouldn't normally think of RSA encryption as a method for authentication. 

> It differs from the standard PKCS#1 OAEP RSA encryption mechanism in that the plaintext is wrapped in a TPM_BOUND_DATA structure before being submitted to the encryption process and that all of the values of the parameters that are passed to a standard CKM_RSA_PKCS_OAEP operation are fixed. On encryption, the version field of the TPM_BOUND_DATA structure must contain 0x01, 0x01, 0x00, 0x00. On decryption, any structure of the
> form 0x01, 0x01, 0xXX, 0xYY may be accepted.

This description of a specific OAEP mechanism seems like it's outside the scope for this issue; it reads to me like the plaintext is just formatted a certain way and then passed directly to OAEP with certain fixed parameters. Then you have to do some extra checks after decryption on the data structure, but if I understand correctly the internal OAEP primitive is not modified and this should be quite possible on top of OT cryptolib's implementation.
 
> https://chromium.googlesource.com/chromiumos/third_party/tpm2/+/569c3c58dc69d5c8628f3c329937c136be38df3f
>
> Besides OAEP, RSAES-PKCS1-v1_5 from https://datatracker.ietf.org/doc/html/rfc3447 is needed for the similar kind of legacy stuff.

The link appears to be about RSA signing; we do already support PKCS#1 v1.5 padding for _signatures_, just not _encryption_. There are no known attacks against the PKCS#1 v1.5 padding for signatures, but the PKCS#1 v1.5 padding method for encryption has been known-vulerable since Bleichenbacher's attack in 1998. I feel strongly that we should not add direct support for it.

**Comment by vsukhoml** [comment]
> Can you provide a little more context for what it is and where it is needed, if possible?

It took some time to find what specifically was used and it happened to be PKCS1.5 padding (not OAEP) with MD5  for the message done by the caller.
Yes, reference to TPM2 & OAEP was misleading - it is an implementation detail that we repurposed TPM2 command instead of using custom vendor command which wouldn't mix in TPM2 specific fields.

> Do I understand correctly that: ...

Correct for all points. The legacy system which use this method is not part of ChromeOS, but we have to support interoperability with it. The purpose to is to actually use RSA Decrypt in TPM with padding done in software to support required RSA signing with MD5/PKCS1.5.  Support for NULL padding enables this use of RSA Decrypt - we can sign with any unsupported algorithms like MD5 or SHA1 if this is the only thing that can work with the other party.

**Comment by jadephilipoom** [comment]
Okay, thanks, I think I understand the concern better now. How about this: we add a new, separate API call to the cryptolib  that runs a bignum modular exponentiation and assumes the exponent is secret. Under the hood this will call the internal RSA modexp routine from `sw/device/lib/crypto/rsa/modexp.h` just like RSA decrypt does. It will not be efficient for encryption/sigverify because it won't take advantage of the non-secret exponent, although it will still be functionally correct for those cases.

I like this solution because it's not the kind of thing someone would call by accident -- I'd worry that simply adding a NULL padding option would lead callers to think that's simply another reasonable padding option they can use. But, for special or legacy use-cases like this, the new function acts as an "escape hatch" that lets you do custom RSA if you know why you're doing it, with appropriate warnings in the docstring. Does that suit your needs @vsukhoml ?

**Comment by vsukhoml** [comment]
Yes, having this `modexp` sounds good. I would appreciate if that solution would accept same format ( `otcrypto_blinded_key_t` ,  `otcrypto_ubblinded_key_t`)  of RSA key to avoid unnecessary changes in storage format as RSA keys are pretty large.

**Comment by johngt** [comment]
@vogelpi / @johannheyszl please note @jadephilipoom analysis above. From cert perspective will this cause an issue? Should this be considered in M3 - otherwise in the schedule it could cause an issue if added in late in the day.

**Comment by johannheyszl** [comment]
@johngt this is a cryptolib issue and does not require any HW changes. IMO it is not M3, and labeled well, thanks
